tests/zquantum/qcbm/cost_function_test.py

```diff
@@ -1,185 +1,191 @@
-            }
-        )
 
-        self.backend = create_object(
+class TestQCBMCostFunction:
+    @pytest.fixture(params=[QCBMCostFunction, create_QCBM_cost_function])
+    def cost_function_factory(self, request):
+        return request.param
+
+    @pytest.fixture(
+        params=[
+            {
+                "distance_measure": compute_clipped_negative_log_likelihood,
+                "distance_measure_parameters": {"epsilon": 1e-6},
+            },
             {
-                "module_name": "zquantum.core.symbolic_simulator",
-                "function_name": "SymbolicSimulator",
-            }
+                "distance_measure": compute_mmd,
+                "distance_measure_parameters": {"sigma": 1},
+            },
+        ]
+    )
+    def distance_measure_kwargs(self, request):
+        return request.param
+
+    def test_evaluate(self, cost_function_factory, distance_measure_kwargs):
+        # Given
+        cost_function = cost_function_factory(
+            ansatz,
+            backend,
+            n_samples,
+            **distance_measure_kwargs,
+            target_bitstring_distribution=target_bitstring_distribution,
         )
 
-        self.gradient_types = ["finite_difference"]
-        self.n_samples = 1
-
-    def test_evaluate(self):
-        for distance_meas, distance_measure_params in param_list:
-            with self.subTest():
-                # Given
-                self.distance_measure = distance_meas
-                distance_measure_parameters = distance_measure_params
-
-                cost_function = QCBMCostFunction(
-                    self.ansatz,
-                    self.backend,
-                    self.n_samples,
-                    self.distance_measure,
-                    distance_measure_parameters,
-                    self.target_bitstring_distribution,
-                )
+        params = np.array([0, 0, 0, 0])
 
-                params = np.array([0, 0, 0, 0])
-
-                # When
-                value_estimate = cost_function(params)
-
-                # Then
-                self.assertEqual(type(value_estimate), ValueEstimate)
-                assert isinstance(value_estimate.value, (np.floating, float))
-
-    def test_evaluate_history(self):
-        for distance_meas, distance_measure_params in param_list:
-            with self.subTest():
-                # Given
-                self.distance_measure = distance_meas
-                distance_measure_parameters = distance_measure_params
-
-                cost_function = QCBMCostFunction(
-                    self.ansatz,
-                    self.backend,
-                    self.n_samples,
-                    self.distance_measure,
-                    distance_measure_parameters,
-                    self.target_bitstring_distribution,
-                )
+        # When
+        value_estimate = cost_function(params)
+
+        # Then
+        assert type(value_estimate) == ValueEstimate
```

Comment:
Wouldn't mypy cover type checks like that automatically for the whole codebase?

Reply:
Yes. I deleted unnecessary checks like this in df1bd95